Matomo statistics provider

snoesberger opened the pull request:
This is a new statistics provider that allows you to access user statistics directly from the Matomo API and display them in the admin UI.
This is my first Opencast module, so I used the InfluxDB statistics provider as a basis and made some adaptations. To avoid unnecessary duplication of Matomo API requests, I introduced the BatchMatomoRequest. Typically, each statistics provider configuration would trigger a Matomo API request. The BatchMatomoRequest caches API responses for statistics provider configurations that use the same API endpoint.


### How to test this patch

You will need a running Matomo instance, and the Paella player must be configured to send user statistics to that instance. Some Paella player user statistics should already be available in Matomo. Instructions on how to configure the Paella player, Matomo, and the statistics provider can be found in the updated documentation section of this PR.


### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

snoesberger committed:
New statistics provider to make user statistics collected with Matomo available within Opencast

Similar to the existing InfluxDB statistics provider, but instead of connecting to an
InfluxDB getting the statistics data from the Matomo reporting API.

snoesberger committed:
Documentation for the new Matomo Statistic Provider

Adding information about the Matomo Statistic Provider to the Administration Guide.

snoesberger committed:
Cache Matomo API responses to avoid duplicate requests

Without this commit each Matomo statistics provider triggers an API request. This can lead to unnecessary
identical API requests in some circumstances. BatchMatomoRequest introduces caching for the Matomo API
requests so that identical requests are sent only once.

wsmirnow commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/package-info.java:
Should this file live here? I guess it should be moved to `src/main/java/org/opencastproject/statistics/provider/`

snoesberger replied:
Ok, I moved the file to the right place.

wsmirnow commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/provider/BatchMatomoRequest.java:
```diff
@@ -0,0 +1,367 @@
+          + "&date="
+          + from.atZone(zoneId).toLocalDate().format(inputFormatter) + ","
+          + to.atZone(zoneId).toLocalDate().format(inputFormatter)
+          + "&period=" + matomoPeriod;
```
Do we expect huge responses here, because `filter_limit=-1`?

snoesberger replied:
Without this parameter, depending on the Matomo config and the API request, it is possible that only a limited number of values is returned. In this case, data exceeding the configured value is returned in one single entry as “others”. To avoid this and to always get all data `filter_limit=-1` is used.

wsmirnow commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/provider/BatchMatomoRequest.java:
```diff
@@ -0,0 +1,367 @@
+
+      String tokenAuth = service.getMatomoApiToken();
+      String requestBody = "token_auth=" + URLEncoder.encode(tokenAuth, StandardCharsets.UTF_8.name());
+      logger.debug("Sending Matomo API requestf or resourceId: {}, method: {}: {}", resourceId, method, url);
```
Typo

```suggestion
      logger.debug("Sending Matomo API request for resourceId: {}, method: {}: {}", resourceId, method, url);
```

snoesberger replied:
fixed

wsmirnow commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/StatisticsProviderMatomoService.java:
```diff
@@ -0,0 +1,175 @@
+  private static final String KEY_MATOMO_API_TOKEN = "matomo.api.token";
+
+  private String matomoApiUrl = "https://your-matomo-instance.org/";
+  private String matomoApiToken = "your-matomo-api-token";
```
Does it make sens to set default values for this configuration? I would prefer not to do that. But then error handling must be improved to handle unset (null) values.

snoesberger replied:
You are right. I removed the default values and improved the error handling. Now, if the Matomo API parameters are not set, error messages are logged. I avoided exceptions to prevent the statistics section of the UI from crashing.

wsmirnow reviewed (changes requested):
Code looks good overall. I will test it on the Testsystem.

geichelberger commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/StatisticsProviderMatomoService.java:
```diff
@@ -0,0 +1,175 @@
+        "service.description=Statistics Provider Matomo Service"
+    }
+)
+public class StatisticsProviderMatomoService implements ManagedService, ArtifactInstaller {
```
Do not use ManagedService, that should not be necessary.
```suggestion
@Component(
    immediate = true,
    service = { ArtifactInstaller.class },
    property = {
        "service.description=Statistics Provider Matomo Service"
    }
)
public class StatisticsProviderMatomoService implements ArtifactInstaller {
```

snoesberger replied:
Thanks for pointing that out! I wasn't aware of that and did copy this part from the InfluxDB statistics provider.

geichelberger commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/StatisticsProviderMatomoService.java:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  @Override
+  public void updated(Dictionary<String, ?> dictionary) {
```
Caused by removing the ManagedService interface.
```suggestion
  @Modified
  public void modified(Map<String, Object> properties) {
```

snoesberger replied:
👍

snoesberger committed:
Implemented suggestions from code review

- moved package-info.java
- better error handling for missing Matomo API parameters
- removed ManagedService which is not necessary in our case

wsmirnow commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/StatisticsProviderMatomoService.java:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  @Deactivate
+  public void deactivate(ComponentContext cc) {
```
Other then `ManagedService`, `modified()` isn't called on service activation but only on configuration updates. Calling it yourself will achieve same behavior.

`ComponentContext` removed because it is unused.

```suggestion
  public void activate(Map<String, Object> properties) {
    logger.info("Activating Statistics Provider Matomo Service");
    modified(properties);
  }

  @Deactivate
  public void deactivate() {
```

snoesberger replied:
Thanks for pointing that out! I added your suggestions to my last commit.

wsmirnow commented on modules/statistics-provider-matomo/src/main/java/org/opencastproject/statistics/provider/matomo/StatisticsProviderMatomoService.java:
```diff
@@ -0,0 +1,178 @@
+import org.opencastproject.util.ConfigurationException;
+
+import org.apache.felix.fileinstall.ArtifactInstaller;
+import org.osgi.service.component.ComponentContext;
```
Unused import.

```suggestion
```

snoesberger replied:
👍

wsmirnow reviewed (changes requested):
Some ToDo's moving from `ManagedService` to annotations.

snoesberger committed:
Call `modified` during service activation

To have the same behavior as with `updated` from  `ManagedService` we need to call `modified` during service activation.
`ComponentContext` isn't needed anymore and can be removed.

wsmirnow reviewed (approved):
Good job. Thank you @snoesberger for the contribution!

gregorydlogan commented:
Merging on the basis of two committers reviewing.
